### PR TITLE
prepare the optimizer

### PR DESCRIPTION
**Pull request by jsteemann**
### Scope & Purpose

Some preparations for the query optimizer, to handle rules that are disabled by default.

- [x] Strictly *new functionality* (i.e. a new feature / new option, no need for porting)
- [x] The behavior in this PR can be (and was) *manually tested* (support / qa / customers can test it)

Enterprise companion PR: https://github.com/arangodb/enterprise/pull/282

### Testing & Verification

This change is already covered by existing tests, such as *all AQL tests*.

https://jenkins01.arangodb.biz/view/PR/job/arangodb-matrix-pr/5404/

### CODE REVIEW

**Comment by mchacki** [comment]
Jenkins after Plan Refactoring:
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/5671/

**Comment by mchacki** [comment]
Jenkins:
http://jenkins.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/5797/

**Comment by mchacki** [comment]
The above linked Jenkins is entirely blue.
Just in case it get's deleted

File: arangod/Aql/ShardLocking.h [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+class Builder;
+}
+
+namespace aql {
+
+struct Collection;
+class ExecutionNode;
+class Query;
+
+class ShardLocking {
```
Comment (danielhlarkin) [inline arangod/Aql/ShardLocking.h]: I know I personally would appreciate some class comments for the new classes, giving a rough idea how the class fits into the overall architecture, how it's used, etc.

File: arangod/Aql/EngineInfoContainerDBServerServerBased.cpp [hunk 2]
```diff
@@ -0,0 +1,527 @@
...
+    }
+  }
+  return localShards;
+}
+
+void EngineInfoContainerDBServerServerBased::TraverserEngineShardLists::serializeIntoBuilder(
+    VPackBuilder& infoBuilder) const {
+  TRI_ASSERT(_hasShard);
+  TRI_ASSERT(infoBuilder.isOpenArray());
+  infoBuilder.openObject();
```
[the remark this answers is not in the document]
Comment (mchacki) [reply arangod/Aql/EngineInfoContainerDBServerServerBased.cpp]: ok, on lengthy functions like this I tend to a agree.

File: arangod/Aql/EngineInfoContainerDBServerServerBased.cpp [hunk 3]
```diff
@@ -0,0 +1,527 @@
...
+ *
+ * @param cc The ClusterComm
+ * @param errorCode error Code to be send to DBServers for logging.
+ * @param dbname Name of the database this query is executed in.
+ * @param queryIds A map of QueryIds of the format: (remoteNodeId:shardId)
+ * -> queryid.
+ */
+void EngineInfoContainerDBServerServerBased::cleanupEngines(
+    std::shared_ptr<ClusterComm> cc, int errorCode, std::string const& dbname,
+    MapRemoteToSnippet& queryIds) const {}
```
[the remark this answers is not in the document]
Comment (mchacki) [reply arangod/Aql/EngineInfoContainerDBServerServerBased.cpp]: Good catch!

File: arangod/Aql/Optimizer.cpp [hunk 4]
```diff
@@ -35,14 +36,68 @@ using namespace arangodb::aql;
...
+  char const* p = name.data();
+  size_t size = name.size();
+
+  if (name[0] == '-') {
+    ++p;
+    --size;
+  }
+
+  if (arangodb::velocypack::StringRef(p, size) == "all") {
+    // enable all rules
```
[the remark this answers is not in the document]
Comment (mchacki) [reply arangod/Aql/Optimizer.cpp]: 0o, that has been there forever

File: arangod/Aql/ShardLocking.h (hunk 1 above)
Comment (mchacki) [reply arangod/Aql/ShardLocking.h]: Agree!

**Comment by jsteemann** [comment]
PR not needed anymore

**Pull request closed without merging**